### ISSUE TITLE
Add to chocolatey or Scoop

### ISSUE THREAD

**Issue by PhilT** [mintty/wsltty#30]
Love this tool (mainly for the truecolor support). It would be great if it was listed in one of the package managers (chocolatey or scoop).

**Comment by mintty** [comment]
No idea what those are. WSL package managers?
Note that wsltty is not a WSL package. Cf. https://github.com/mintty/wsltty/issues/20#issuecomment-274791898

**Comment by PhilT** [comment]
These are Windows package managers. They allow you to easily install and update software on the OS via the commandline. Linux and MacOS has had these for years. Windows is finally catching on. With chocolatey instead of having to download individual packages you can simply do `cup --all -y` and it'll install the latest versions of any programs managed by chocolatey. I recommend checking it out: https://chocolatey.org/

**Comment by mintty** [comment]
You are welcome to build a wsltty chocolate package, register it there, or whatever is needed to facilitate it.
I'll mention it on the wsltty home page if it works.

**Comment by PhilT** [comment]
If I get the chance I'll look into at the weekend.

**Comment by Link-Satonaka** [comment]
For a package to work in chocolatey, the installer needs to support running in a silent mode.

I don't think your installer currently supports silent install, which I think should be the /Q

`wsltty-0.7.5-install.exe /Q`

**Comment by Link-Satonaka** [comment]
Installing this to chocolatey might be a bit more complicated than your conventional package- in fact I owe a few documentation commits to the choco docs explaining why.

In short, you can't install things to APPDATA, because APPDATA is a folder specific to the current user, and chocolatey is expected to be run as an admin (root) user. In Linux parlance- you're installing to `$HOME/.local/wsltty`, but since chocolatey is run as root, it ends up in `/root/.local/wsltty`, inaccessible to the end user.

A package like this (one which doesn't install to Program Files) would need to be installed to `%ChocolateyInstall%\lib\wsltty\tools`, which typically translates to `%ProgramData%\chocolatey\lib\wsltty\tools`. Everyone has read (but not write) access to this folder- it's the equivalent of `/usr/bin`.

Which, of course, spells trouble for the mintty's config, which appears to use the current working directory by default. A program's config would usually go into `Appdata\Roaming\wsltty` (`$HOME/.config/wsltty`). Programs which absolutely need to write to the working directory get installed to their own special folder, usually `C:\tools\wsltty`, where everyone has full read/write access.

That `C:\tools` folder is a last resort- we don't want to use it if we don't have to. Luckily, You provide the -c option, but it won't work if the specified file doesn't exist.

The wsl batch script will need to be adjusted to 

- Point to the new location of the bin folder
- Detect if `%APPDATA%\wsltty\.minttyrc` exists, create it if not
- Run mintty pointing at the above config file

I can do this, I just wanted you to be aware that the chocolatey package will not be an exact 1:1 install of the upstream installer.

**Comment by mintty** [comment]
Let's sort out the installation and configuration folders, clarifying some confusion:
* The wsltty package installs into %LOCALAPPDATA% (not %APPDATA%). While I agree this is a weird place to install, it's the same place that WSL is installed to, thus the choice. I do not think a common terminal package that's an add-on to a user-specific system installation should itself be installed in a common location. But if you think there should be a package to install into $PROGRAMFILES, go ahead. Installation into the domain of any third-party software, however, does not seem appropriate to me.
* The mintty configuration file can be located in various places. $HOME/.minttyrc and /etc/minttyrc are two options for the single configuration file, $HOME/.config/mintty (not wsltty) is an option for the config folder; all of these refer to the installation root of the mintty-embedding standalone cygwin package, which would automatically move if you install wsltty anywhere else. The final option, $APPDATA/mintty/config, is an option to provide common configuration for independent mintty installations, and does not need to be referred explicitly, it's coded within mintty; I don't think the folder should be created automatically.

**Comment by Link-Satonaka** [comment]
> The wsltty package installs into %LOCALAPPDATA% (not %APPDATA%)

I know- but since with Chocolatey we will not be using either for the binary, I made the (unstated) decision that config files would go in regular %APPDATA%.

> Installation into the domain of any third-party software, however, does not seem appropriate to me.

When in Rome... `%ChocolateyInstall%\lib\wsltty\tools` is where wsltty would go as a Chocolatey package. To create a Chocolatey package and defy Chocolatey conventions because it "does not seem appropriate" would be inappropriate. Chocolatey is the way it is for good reason- if you go to the developer gitter, they will happily explain to you the challenges involved with wrangling the many wonky kinds of Windows software, much better than I can.

For my part, I can explain the aspect I am most familiar with: user vs admin. Chocolatey has (still in progress) support for user-only mode, where admin/root privileges are not needed. Under this setup, `%ChocolateyInstall%` refers to a user writable location. Most packages have two versions: {pkg}.install and {pkg}.portable. There are a few different implications for each, but the most basic separation is that {pkg}.install inherently requires admin privileges to work, and {pkg}.portable does not.

For wsltty, there is no admin requirement- the default thing to do is treat it like we would treat a portable package. Installing to `%ChocolateyInstall%` is a simple way to make one single package function under both admin and user Chocolatey instances.

In proposing that I install wsltty into Program Files, the end result will be that I create two packages: wsltty.install and wsltty.portable, where the portable will go into `%ChocolateyInstall%`. The install package has no purpose, and it doesn't quite make sense to create an install package for the sole reason of avoiding `%ChocolateyInstall%` ;)

> The final option, $APPDATA/mintty/config, is an option to provide common configuration for independent mintty installations, and does not need to be referred explicitly, it's coded within mintty; I don't think the folder should be created automatically.

In my limited testing, moving wsltty just caused the configuration file to move with the binary. If no config exists on the system and mintty is moved to a read only location, how will it respond if the user presses "save"?

**Comment by mintty** [comment]
I'm not familiar with Chocolatey but thanks to your response I'm getting the point.
So well, I agree that you may proceed as indicated, but the config file to be created would be %APPDATA%\mintty\config (not %APPDATA%\wsltty\.minttyrc) to refer to the common config place already in use. If you want to create a specific one, I'd still suggest to agree on something in %LOCALAPPDATA%, not %APPDATA%, if only to avoid confusion with the "common config" option. (Or maybe %APPDATA%\wsltty... as you suggest but then as a folder, so the other config sub-folders can also be hosted there; this will need some update in mintty itself to be addressable.)

**Comment by Link-Satonaka** [comment]
OK, so all I need to add to the batch file is (the batch equivalent of) `touch %APPDATA%\mintty\config`, and mintty will find it without specifying `-c`? Cool.

Let me know if you'd prefer somewhere else.

**Comment by mintty** [comment]
Yes.
Well, some people may like to use the common config feature, while others might prefer a WSL-specific dedicated config, so I'm not sure...

**Comment by Link-Satonaka** [comment]
I'm not in any rush, you can get  back to me whenever.

Here are the available options I can think of:

1. I simply don't create or otherwise specify a config, and use `%ChocolateyInstall%`. When pressing 'save', mintty will throw an error, because it can't write to the working directory. Users will always need to manually create a config file.
2. Mintty source code is updated to have a fallback save location when the working directory is not writable.
3. I create a config somewhere.
4. I install mintty to "ChocolateyToolsLocation" (usually `c:\tools\wsltty`). This is completely user writable, but it's generally reserved for programs which *require* write access to the working directory. Furthermore, if mintty prioritizes workingdirectory config over %APPDATA% config, then using this means there will likely be one config file for all wsltty users on the PC.

Correct me if I'm wrong, but it has been my experience that mintty always defaults to a local config, and the user-wide `%APPDATA%\mintty\config` is an opt in; the user must manually create it since afaik at no point is it automatically created. If all this is true, it would be internally consistent to by default make a wsltty specific config, perhaps at `%APPDATA%\wsltty\config`.

**Comment by mintty** [comment]
Please try the preferred installation target (%ChocolateyInstall%) and test whether everything works fine with configuration. For a dedicated config location, I would however suggest to consider %LOCALAPPDATA%\wsltty\config, because %LOCALAPPDATA% is the location chosen by WSL for itself.

**Comment by Link-Satonaka** [comment]
Minor issue: language files are currently found in `%LOCALAPPDATA%\home\%USERNAME%\.config\mintty\lang`, but I can't create that folder programatically, unless I have the launching batch files copy them every time you start wsltty. Is there any location the language files could go where they'd be available as a part of the software, not the per user config?

**Comment by mintty** [comment]
You mean you cannot create them from within WSL? Probably an access rights issue.
Why copy every time? Are they getting deleted?

We could create them during installation and try to make them writable for everyone.
Or you could use one of the alternative locations:
 %LOCALAPPDATA%\usr\share\mintty\lang
 %APPDATA%\mintty\lang

**Comment by Link-Satonaka** [comment]
I'm not sure how I would go about creating those files within WSL. If this is something the user can be expected to do themselves, would it be acceptable to just leave out the language files?

I can't use any of those proposed locations, because thinking in terms of a system-wide install, I cannot place things in C:\Users, much like a Linux package manager will never touch $HOME.

To get around this, I could alter wsl.bat such that it attempts to copy or symlink the lang files from the systemwide location to the user profile.

That gets sticky, though. Symlinks require administrative privileges, hardlinks don't work across drives/partitions. Copy every time because if you ever update the upstream files, only copying once means the user will never see updated versions of those files, nor new additional languages.

I might try creating an NTFS junction in wsl.bat

**Comment by Link-Satonaka** [comment]
Here's what I've come up with: https://github.com/Link-Satonaka/chocopkgs/blob/db5421bea0c3cb0bbd82df16f6b0bf68619b3128/wsltty/tools/chocolateyinstall.ps1

`$toolsdir` ends up being `C:\ProgramData\chocolatey\lib\wsltty\tools`, and the wsltty installer is extracted to this directory as if it were a zip archive.

Lines 12-19 are just recreating the directory structure your installer would have created

22 Extracts `po.zoo` into `$toolsdir\lang`

25-27 edits the three batch files sich that they point to the correct location

30-32 adds the `--config "%LOCALAPPDATA%\wsltty\config"` to the batch files

35-44 merges each batch file with https://github.com/Link-Satonaka/chocopkgs/blob/1dc523c26ff344bd704f3e335fa326e8373b95a1/wsltty/tools/create-dirs.bat

47-49 creates "shims" in `%ChocolateyInstall\bin%`; effectively this step is just making the batch files available on PATH.

------

I initially tried `%LOCALAPPDATA%\usr\share\mintty\lang` and it did not work, so I switched to `
%APPDATA%\mintty\lang`. Let me know if the former location should work, so I can keep all the files within %LOCALAPPDATA%.

**Comment by mintty** [comment]
What about the desktop shortcuts?

If you install somewhere else, the implicit cygwin root is changed and the mintty reference change with it, so %LOCALAPPDATA% is irrelevant now. You redirected back the config file with an explicit --config option, but such an option is not available for the whole config folder. This is a gap that may be closed in mintty for this purpose.

**Comment by mintty** [comment]
I notice MKLINK which is not a standard Windows tool. Would this work without additional dependencies?
But we'll probably do without it anyway after sorting out the config reference issue.

**Comment by Link-Satonaka** [comment]
I believe MKLINK has been part of the standard Windows toolset since Vista, though I agree it would be better to not have to use it. What specifically do you intend to do? From my perspective, there are files included with the installer, and files generated by the program upon launch- typically the former reside in the program directory, and the latter in userspace. Since `lang` is installed and not generated, it would help me if `lang` was detected by mintty within the program directory, rather than userspace. I can't see any other way of avoiding MKLINK, unless we decide to use `c:\tools\wsltty` after all, keeping in mind that is still system wide install and all users would share the config (unless? each user is careful to not create a config in `c:\tools\wsltty` and manually create their configs in %APPDATA%? I'm a unclear on that point).

I'll add desktop shortcuts if you want them- though I personally dislike it when programs keep cluttering my desktop, and even even I clean it, the clutter comes back every time the program updates ;)

Start Menu shortcuts I left out as well- I can add them, but since the batchfiles are shimmed on PATH they will appear in the Start Menu's searchbox.

Let me know

**Comment by mintty** [comment]
Ah right, mklink is a cmd.exe built-in. However, creating symbolic links needs special privileges in Windows, so better to be avoided. I see another problem:
```
MKLINK /J "%APPDATA%\mintty\lang" "%ChocolateyInstall%\lib\wsltty\tools\lang"
```
If you do this because %ChocolateyInstall% is not writable, I don't think it will work because using a symbolic link, you would still need a writable target directory (by POSIX means; would be surprised if Windows links are capable of overriding access rights).

Desktop shortcuts are the same files as Start menu shortcuts, they need to be created with the proper target. If you use a variable (%ChocolateyInstall%) and the variable is set in the Windows environment, the shortcut can be taken as provided with the package, without modification. (That's what my package does.)
Even if you don't install desktop shortcuts (personally, I don't like them either), they should be installed in the Start menu (like my package does) so people can copy them to their desktop as they like. People typically want shortcuts.

Third consideration: Your setup now seems to be using completely different locations for mintty configuration ("%LOCALAPPDATA%\wsltty\config" and "%ChocolateyInstall%\lib\wsltty\tools\lang"). This isn't really convincing; I'll add a command line option to set the whole config dir.

**Comment by Link-Satonaka** [comment]
This is an NTFS Junction, which does not need any special privileges, nor does the location need to be writable for the junction to work.

OK, I'll add start menu shortcuts.
> Third consideration: Your setup now seems to be using completely different locations for mintty configuration ("%LOCALAPPDATA%\wsltty\config" and "%ChocolateyInstall%\lib\wsltty\tools\lang"). This isn't really convincing; I'll add a command line option to set the whole config dir.

I'm not sure if we're understanding each other. You seem to be saying that the lang files are part of a config, rather than a general program asset that would benefit from being available in a systemwide install. I think mintty would benefit from looking for lang files locally, such that they will be found **independent** of user config. Lang would be located in the same way that the file `bin\wslbridge-backend` is found within the bin folder.

It's important to note that I have placed lang within `"%ChocolateyInstall%\lib\wsltty\tools\lang` because I _have_ to. I cannot, as a part of the installer package, create a folder in %APPDATA% nor %LOCALAPPDATA%. If you are of the opinion that lang must exist in user config space, the only thing I can do is abuse your batch files to do one of two things:

1. junction the lang files into the user config directory , every time wsltty is run
2. copy the lang files into the user config directory, every time wsltty is run

(This means any start menu shortcuts I create will also point to the batch files, not mintty.exe like your shortcuts do)

Being able to specify a config directory does not change any of the above. Understand that, even if you add a config directory, **I still need to rework your batch files to copy the lang files into the config folder.**

If you are going to make changes to accommodate the way this package works, here are a few options I can think of:

1. lang needs to be a local resource, not treated as a config item
2. `--lang-dir` could be a thing?
3. mintty itself extracts a locally found po.zoo into the specified `--config-dir`,- similar to how .minttyrc is created automatically if it does not already exist.
4. other???

**Comment by mintty** [comment]
> lang needs to be a local resource, not treated as a config item

Not sure what you mean. You can `-o Language=...`.

> --lang-dir could be a thing?

Too specific; we can talk about `--config-dir`.

> mintty itself extracts a locally found po.zoo into the specified --config-dir,- similar to how .minttyrc is created automatically if it does not already exist.

I don't see the benefit, and mintty would have to be bundled with zoo.exe.

> c:\tools\wsltty and the implication of only one config unless specific care is taken by each user

Weren't you planning for %ChocolateyInstall% rather than C:\tools? What do you mean by one config? The --config-dir?

**Comment by Link-Satonaka** [comment]
Let me try to explain the problem again.

- Normally, an installer includes all the files the program needs to run.
- Normally, **all of the files within the installer** are simply installed to the read-only `Program Files`.
- The program itself, **not the installer**, is what creates configuration items; this is because only the program itself has access to user space, the installer does not.

With this, any user on the system can run the software, and each user will have their own config file.

wsltty/mintty breaks this formula by

1. Expecting the installer to provide lang files
2. Expecting the lang files to be in user $HOME

For a proper installation to work, **you cannot have both 1 and 2!!**. This is not a concept unique to Windows- in Linux, too, it is forbidden for a package manager to touch your $HOME. Are you a Linux user? Think about trying to install wsltty/mintty into Linux- you're trying to get a package manager to put files in `$HOME/.config/wsltty`. This isn't something you're allowed to do, and it's no different here in Windows.

Moving forward, I can only think of two options:

1. You do nothing, wsltty/mintty stays exactly as it is. Chocolatey package puts it in `c:\tools\wsltty`.
2. wsltty/mintty is updated to honor the separation of system files and user $HOME.

>Weren't you planning for %ChocolateyInstall% rather than C:\tools? What do you mean by one config? The --config-dir?

If we choose to do 1, I will only update location of the binary within the batch files. I will not point at a `--config-dir`. Mintty will automatically create `c:\tools\wsltty\.minttyrc`, and consequently will be the config file that gets used. It's up to you how much you think that actually matters.

**Comment by mintty** [comment]
Answering in reverse order:

C:\tools is a global place, but .minttyrc is a per-user resource, therefore this does not look proper for me.

For the comparison with other installation procedures, If I normally install a package for myself, not system-wide, the packager manager would have access to my personal folders.

And the choice to install completely into user space may be a weird one, agreed, but the point is it's the same choice that WSL takes, so I don't see why a support package installed in parallel should do something different.

So there are points against both your options 1 and 2, and we'll have to check further...
But maybe, since WSL is still in beta state, they are eventually going to change their installation strategy? In that case I would revise mine as well, of course.

**Comment by Link-Satonaka** [comment]
> For the comparison with other installation procedures, if I normally install a package for myself, not system-wide, the packager manager would have access to my personal folders.

An individual installer is bound to no rules, which is exactly why we are having this conversation: the complete lack of enforced packaging standards for software in Windows == chaos.

While chocolatey tries to accommodate all different kinds of installers, there is only so much I can do with what is given to me. In this case, it is not a matter of "I think there are reasons to install to user directories"- I literally cannot do it.

**Comment by mintty** [comment]
OK, this was only to explain the current choice and why I wouldn't like to spend effort to change it at this time. Of course you are welcome to create a different installation package from wsltty, or a suitable wrapper. I will also support that if we agree on a reasonable arrangement.

My other point is that I'm puzzled about the proposal of a global folder to include config files. Those must be located in a user-specific folder. A global place may provide default/fall-back resource like language and theme files, but for them to be adapted and especially for the main config file, a user-specific place must exist.

**Comment by Link-Satonaka** [comment]
>I'm puzzled about the proposal of a global folder to include config files

I was just trying to give you a complete list of options. The `C:\tools` option is least invasive from my end; it ONLY changes the installation destination (away from %APPDATA%).

> ...config files. Those must be located in a user-specific folder

Agreed; and typically the software itself creates those config files, not the installer. Mintty does create a .minttyrc automatically, but only in the current working directory, which isn't useful here.

>A global place may provide default/fall-back resources like language and theme files

This would be very useful to me!

**Comment by mintty** [comment]
Short from the next release, after implementing --configdir and some other configuration details, one issue occurs to me: if you install to a read-only directory hierarchy, /tmp is not usable either, so mintty downloads (for version checking and for theme downloads) need to got somewhere else. That will probably be the last change for mintty 2.7.7.
I then plan to add 2 optional parameters to the install script, for the installation target dir and the configuration dir, and inject them into the installed wsl*.bat files. To inject them also into the desktop shortcuts, the cygwin tool mkshortcut could be added, along with some library dependencies; still wondering whether a more light-weight solution is available.

**Comment by mintty** [comment]
Released 0.7.7.

**Comment by mintty** [comment]
I had overlooked to adapt the installation and configuration description on the wsltty home page to the changed targets and script options. Fixed just now.

**Comment by mintty** [comment]
Just curious: Since I refactored deployment to support dedicated and separate installation of software and configuration, does anyone still plan to build a 3rd-party installer around it?

**Comment by Link-Satonaka** [comment]
Yes, I'll still work on the chocolatey package :)

**Comment by lucastheisen** [comment]
I cant find the `chocolatey` package, did this ever happen?  Is it deployed as something else?

**Comment by mintty** [comment]
If it happens, it will be notified here.

**Comment by AeliusSaionji** [comment]
OK, I took another stab at this @mintty and as I suspected, we still have a miscommunication.

`install.bat customlocation` works fine, but no shortcuts are created. I see you have a `configure WSL shortcuts.lnk` which runs an sh script through a provided dash.exe, and my guess is that this is intended to generate shortcuts.

However, it expects the existence of `%LOCALAPPDATA%`, which I cannot use. In order for this to work, files need to be created in `%LOCALAPPDATA%` for every single user, and the only way that is going to happen is if some some component within wsltty creates those files on the user's first run.

The easy way out is to write another script which can copy the appropriate files to `%LOCALAPPDATA%`, and I'll create a global shortcut in the start menu, so all users will have a shortcut which will generate shortcuts + place the correct files locally for them. There's probably a better way. I didn't look to deeply into the inner workings this time.

**Comment by AeliusSaionji** [comment]
idk, maybe I've misunderstood something here...
looking into `install.bat`

would it be correct to say that each user is supposed to run install.bat?

I'll continue playing with it tomorrow.

**Comment by mintty** [comment]
As you're taking this up again, I'll reopen the issue.
The installer concept was, as WSL distributions themselves are installed per user, wsltty does the same.
The current script install.bat deploys resources into %installdir% and prepares %configdir% for the user's own resource configuration. Both are in %LOCALAPPDATA% by default but both can be changed. So there is no problem if you deploy resources in a common location and maybe don't touch the user's config at all, so there will be none. Whether a user config area should be prepared, maybe on first invocation (perhaps better than an explicit script), is up to discussion.

**Comment by AeliusSaionji** [comment]
OK, I understand now. Thanks!

This package is pretty simple. It does two things:
- Extracts contents of installer to `%ChocolateyInstall%\lib\wsltty\tools\wslttyinstall`
- Creates a shortcut `%ALLUSERSPROFILE%\Microsoft\Windows\Start Menu\Programs\WSL Generate Shortcuts.lnk` which simply links to wsltty's `install.bat`.


Let me know if you have any suggestions/complains. @lucastheisen @PhilT 
https://chocolatey.org/packages/wsltty

**Comment by mintty** [comment]
I _think_ I installed Chocolatey before, not sure though. As I'm reluctant to read all the instructions, can you please quote precise instructions how to install it, and how to install the wsltty package with it?

**Comment by AeliusSaionji** [comment]
Sure thing.

First, run cmd *as administrator* (Start > cmd > right click > run as administrator)

Then paste this snippet. This is taken from https://chocolatey.org/install if you want to copy from there rather than here.

```cmd
@"%SystemRoot%\System32\WindowsPowerShell\v1.0\powershell.exe" -NoProfile -InputFormat None -ExecutionPolicy Bypass -Command "iex ((New-Object System.Net.WebClient).DownloadString('https://chocolatey.org/install.ps1'))" && SET "PATH=%PATH%;%ALLUSERSPROFILE%\chocolatey\bin"
```

After successful install, run powershell as administrator. Chocolatey always requires an administrative shell at this point.

From powershell, run `choco install wsltty --version 1.8.4`. Done!

**Comment by mintty** [comment]
Released to 1.8.5, with some significant compatiblity improvement (in case WSL mount points are reconfigured). Once you update the choco package to 1.8.5, I'll add a hint about this installation option to the project page.

**Comment by mintty** [comment]
Added section to the project page.
Actually, 1.8.5.2 was only a workaround attempt for virus scanner problems (it's not a separate release); it has the drawback that it cannot install desktop shortcuts for specific WSL distributions (only 1 for the default distro).
1.8.5 would be the right up-to-date package for now, or wait for 1.8.6 to come soon.

**Comment by AeliusSaionji** [comment]
Ah, oops my bad. I wondered why it wasn't its own release.

**Comment by mintty** [comment]
Released 1.9.0.

**Comment by AeliusSaionji** [comment]
I have AppVeyor running my update scripts every 12 hours. As long as releases continue to follow the pattern `https://github.com/mintty/wsltty/releases/download/$version/wsltty-$version-install.exe`, the chocolatey package will autonomously stay up to date 😉
